### PR TITLE
Eliminate Python 2.7 support on buildbot master only

### LINKED ISSUE #4439: Drop Python 2.7 support for master.
## Reasonings:

- Python 2 is going unmaintained as soon as 2020: https://pythonclock.org/

- Maintaining compatibility for both python2 and python3 is expensive resource wise

- Will allow to use python3 features like async/await, which adds a big performance boost for asynchronous apps like Buildbot

## Timeline

Buildbot 2.0 is planned to be released on February 2019.
This version will drop support for 2.7 in the master.

Buildbot 1.7 and 1.8 will contain loud warning for python2 users at Buildbot start to invite them to upgrade.

Original issue:
```
There have been some talks about no longer supporting running master under python 2.x 
some time in the future.
 As 2.0 version will be released soon, maybe it would make sense to mark python 2.x support 
deprecated so that it can be removed in the next major release - 3.0?
```

### PR DESCRIPTION
**Pull request by rodrigc**
Update documentation, installer, and CI to remove support for Python 2.7 from the buildbot master.

Fixes #4439

### CODE REVIEW

**Comment by p12tic** [comment]
I think there were some discussions of bumping to python 3.5 too, weren't there? @tardyp did you make some decision in this regard? 

One of the benefits would be possibility of using async/await at least internally, even if that's not exposed in the APIs i.e. we wrap async functions with inlineCallbacks. I believe there are significant optimizations possible, but I haven't had time to explore that yet. OTOH, I already have a non-polished PR that improves the performance of generator+inlineCallbacks-based twisted code by up to 2 times https://github.com/twisted/twisted/pull/1090.

**Comment by tardyp** [comment]
I would merge the part of this PR that are changing the default to python3.

but we need to continue test and support python2 at least for a few months as you suggested yourself.

File: .bbtravis.yml [hunk 1]
```diff
@@ -25,7 +25,7 @@ env:
   - TWISTED=latest SQLALCHEMY=latest TESTS=js HYPER_SIZE=m1
   - TWISTED=latest SQLALCHEMY=latest TESTS=smokes HYPER_SIZE=m2
 
-  - TWISTED=16.1.0 SQLALCHEMY=latest TESTS=trial
+  - TWISTED=17.9.0 SQLALCHEMY=latest TESTS=trial
```
Comment (tardyp) [inline .bbtravis.yml]: I think TWISTED 17.9 is generating those StopIteration warnings

**Comment by rodrigc** [comment]
@p12tic can you help me fix the tests that depend on dictionary ordering?  You can send pull requests directly to my branch.

**Comment by tardyp** [comment]
restarted build with new docker  image from your PR

File: master/buildbot/db/sourcestamps.py [hunk 2]
```diff
@@ -166,7 +167,27 @@ def _rowToSsdict_thd(self, conn, row):
                 ssdict['patch_subdir'] = row.subdir
                 ssdict['patch_author'] = row.patch_author
                 ssdict['patch_comment'] = row.patch_comment
-                body = base64.b64decode(row.patch_base64)
+
```
Comment (tardyp) [inline master/buildbot/db/sourcestamps.py]: can you explain what you are trying to do, what the problem you are trying to fix?

Comment (rodrigc) [reply master/buildbot/db/sourcestamps.py]: @tardyp there are some problems in `TestRealDB` : https://travis-ci.org/buildbot/buildbot/jobs/469199833#L1183

when tested against a real MySQL or Postgres database.

I was trying to chase these down.

File: .travis.yml [hunk 3]
```diff
@@ -19,7 +19,7 @@ language: python
 # Available Python versions:
 # http://about.travis-ci.org/docs/user/ci-environment/#Python-VM-images
 python:
-  - "2.7"
+  - "3.7-dev"
```
Comment (yan12125) [inline .travis.yml]: On trusty, 3.7-dev points to an ancient commit before 3.7.0, as [OpenSSL on trusty is too old](https://github.com/travis-ci/travis-ci/issues/9069):

> Python 3.7.0a4+

(https://travis-ci.org/buildbot/buildbot/jobs/474355722#L562)

I guess you want the latest Python 3.7? Then you need to change `dist: trusty` to `dist: xenial`

Comment (rodrigc) [reply .travis.yml]: OK, I switched it to Python 3.7.1 on xenial

File: master/buildbot/db/sourcestamps.py (hunk 2 above)
Comment (rodrigc) [reply master/buildbot/db/sourcestamps.py]: I found and fixed the issue here: https://github.com/buildbot/buildbot/pull/4500

File: .bbtravis.yml (hunk 1 above)
Comment (rodrigc) [reply .bbtravis.yml]: No, the  StopIteration warnings were being caused by sqlalchemy 0.8.0.
I fixed those warnings here: https://github.com/buildbot/buildbot/pull/4499

**Comment by tardyp** [comment]
that will be a lot to rebase.. did you apply some kind of script for that?

**Comment by rodrigc** [comment]
@tardyp No, I did it manually.  I will maintain the branch and try to cherry-pick some fixes from it to master to make it easier.

**Comment by tardyp** [comment]
@rodrigc not sure why circleci is failing. did you look at it already?

**Comment by rodrigc** [comment]
@tardyp Looks like we need to add an explicit ignore of:
```
builtins.PendingDeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
```

**Comment by rodrigc** [comment]
not sure, maybe there is some 3rd party dependency that got pulled in recently.  It wasn't happening on this branch before.

**Comment by rodrigc** [comment]
Here is a successful CircleCI build on this branch before:
https://circleci.com/gh/buildbot/buildbot/7442

**Comment by rodrigc** [comment]
@tardyp I got all the CI's passing with this branch now, after adding https://github.com/buildbot/buildbot/pull/4470/commits/5ae18d2e385e85aa91440ef30fc45a589a3140db

**Comment by tardyp** [comment]
@rodrigc could you please rebase?

**Pull request merged**